### DISCUSSION TITLE
Ability to use sdf functions in canvas_item shaders outside of fragment() and light(), e.g. in helper functions

### DISCUSSION THREAD

**Discussion by demolen** [godotengine/godot-proposals#10047] [2D]
### Describe the project you are working on

Any game

### Describe the problem or limitation you are having in your project

it's not possible to use sdf functions outside of light and fragment parts, this forces user not to use helper functions and write most of their code inside fragment or light or pass such values as parameters. All of this result in "dirty" code that is not well readable.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

allow usage of sdf_functions outside of fragment and light.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

--

### If this enhancement will not be used often, can it be worked around with a few lines of script?

--

### Is there a reason why this should be core and not an add-on in the asset library?

--

**Comment by AThousandShips** [comment]
Converted to a discussion as there were no concrete steps for implementing or achieving this

**Comment by Calinou** [comment]
cc @Chaosus 

This might be possible if the code that forbids calling SDF functions outside `fragment()` and `light()` is updated to use https://github.com/godotengine/godot/pull/93358, which takes the function's callers into account for filtering. See https://github.com/godotengine/godot/pull/93590 for an example.

**Reply by Chaosus** [reply]
Ah, yes, this might be a good idea...

**Reply by Chaosus** [reply]
Though they pass like stage functions, not sure if It's easy to make it use in global space of custom functions.

**Comment by Chaosus** [comment]
I've created a PR to fix this: https://github.com/godotengine/godot/pull/93650

**Reply by demolen** [reply]
Well done and thank you Chaosus!

**Reply by demolen** [reply]
> I've created a PR to fix this: [godotengine/godot#93650](https://github.com/godotengine/godot/pull/93650)

Now that you're on it. I would also suggest adding the ability to use UV, SCREEN_UV, etc... inside the custom functions as well. should I open a separate discussion for that?

**Reply by Chaosus** [reply]
That would be harder than ever because then we need to check which built-in can be referenced in one processor function and prevent mixing them with another in whole call hierarchy. I would suggest to wait until the previous PR's will be merged.

**Reply by Chaosus** [reply]
Plus some of the built-in are impossible to be transferred to functions because they are not global and declared inside a main processor function.
